File: core/src/engine/kana_parser.cpp
#include "kana_parser.h"

#include <algorithm>
#include <optional>
#include <stdexcept>

#include "mora_list.h"

namespace voicevox::core::engine {
static const std::map<std::string, MoraModel> text2mora_with_unvoice() {
  std::map<std::string, MoraModel> text2mora_with_unvoice;
  const std::string* mora_list = mora_list_minimum.data();
  int count = 0;
  while (count < mora_list_minimum.size()) {
    std::string text = *mora_list;
    std::optional<std::string> consonant = *(mora_list + 1);
    std::optional<float> consonant_length;
    if (consonant->empty()) {
      consonant = std::nullopt;
      consonant_length = std::nullopt;
    } else {
      consonant_length = 0.0f;
    }

    std::string vowel = *(mora_list + 2);

    MoraModel mora = {text, consonant, consonant_length, vowel, 0.0f, 0.0f};

    text2mora_with_unvoice[text] = mora;

    if (vowel == "a" || vowel == "i" || vowel == "u" || vowel == "e" || vowel == "o") {
      std::string upper_vowel = vowel;
      std::transform(upper_vowel.begin(), upper_vowel.end(), upper_vowel.begin(), ::toupper);
      MoraModel unvoice_mora = {text, consonant, consonant_length, upper_vowel, 0.0f, 0.0f};

      text2mora_with_unvoice[UNVOICE_SYMBOL + text] = unvoice_mora;
    }

    mora_list += 3;
    count += 3;
  }

  return text2mora_with_unvoice;
}

template <typename N>
std::string extract_one_character(const std::string& text, N pos, N* size) {
  // UTF-8の文字は可変長なので、leadの値で長さを判別する
  unsigned char lead = text[pos];

  if (lead < 0x80) {
    *size = 1;
  } else if (lead < 0xE0) {
    *size = 2;
  } else if (lead < 0xF0) {
    *size = 3;
  } else {
    *size = 4;
  }

  return text.substr(pos, *size);
}

AccentPhraseModel text_to_accent_phrase(std::string phrase) {
  std::optional<unsigned int> accent_index = std::nullopt;

  std::vector<MoraModel> moras;

  int base_index = 0;
  std::string stack = "";
  std::optional<std::string> matched_text = std::nullopt;

  const std::map<std::string, MoraModel> text2mora = text2mora_with_unvoice();

  int outer_loop = 0;
  while (base_index < phrase.size()) {
    outer_loop++;
    int char_size;
    std::string letter = extract_one_character(phrase, base_index, &char_size);
    if (letter == ACCENT_SYMBOL) {
      if (moras.empty()) {
        throw std::runtime_error("accent cannot be set at beginning of accent phrase: " + phrase);
      }
      if (accent_index != std::nullopt) {
        throw std::runtime_error("second accent cannot be set at an accent phrase: " + phrase);
      }

      accent_index = moras.size();
      base_index += char_size;
      continue;
    }
    int watch_char_size;
    for (int watch_index = base_index; watch_index < phrase.size(); watch_index += watch_char_size) {
      std::string watch_letter = extract_one_character(phrase, watch_index, &watch_char_size);
      if (watch_letter == ACCENT_SYMBOL) break;
      stack += watch_letter;
      if (text2mora.find(stack) != text2mora.end()) {
        matched_text = stack;
      }
    }
    if (matched_text == std::nullopt) {
      throw std::runtime_error("unknown text in accent phrase: " + stack);
    } else {
      moras.push_back(text2mora.at(*matched_text));
      base_index += matched_text->size();
      stack = "";
      matched_text = std::nullopt;
    }
    if (outer_loop > LOOP_LIMIT) throw std::runtime_error("detect infinity loop!");
  }
  if (accent_index == std::nullopt) throw std::runtime_error("accent not found in accent phrase: " + phrase);

  AccentPhraseModel accent_phrase = {
      moras,
      static_cast<unsigned int>(*accent_index),
  };
  return accent_phrase;
}

std::vector<AccentPhraseModel> parse_kana(std::string text) {
  std::vector<AccentPhraseModel> parsed_results;

  std::string phrase = "";

  size_t char_size;
  for (size_t pos = 0; pos <= text.size(); pos += char_size) {
    std::string letter;
    if (pos != text.size()) {
      letter = extract_one_character(text, pos, &char_size);
    }
    phrase += letter;
    if (pos == text.size() || letter == PAUSE_DELIMITER || letter == NOPAUSE_DELIMITER) {
      if (phrase.empty()) {
        throw std::runtime_error("accent phrase at position of " + std::to_string(parsed_results.size() + 1) +
                                 " is empty");
      }
      bool is_interrogative = phrase.find(WIDE_INTERROGATION_MARK) != std::string::npos;
      if (is_interrogative) {
        if (phrase.find(WIDE_INTERROGATION_MARK) != phrase.length() - char_size) {
          throw std::runtime_error("interrogative mark cannot be set at not end of accent phrase: " + phrase);
        }
        phrase = phrase.replace(phrase.length() - char_size, char_size, "");
      }
      AccentPhraseModel accent_phrase = text_to_accent_phrase(phrase);
      if (pos < text.size() && letter == PAUSE_DELIMITER) {
        MoraModel pause_mora = {PAUSE_DELIMITER, std::nullopt, std::nullopt, "pau", 0.0f, 0.0f};

        accent_phrase.pause_mora = pause_mora;
      }
      accent_phrase.is_interrogative = is_interrogative;
      parsed_results.push_back(accent_phrase);
      phrase = "";
    }
  }
  return parsed_results;
}

std::string create_kana(std::vector<AccentPhraseModel> accent_phrases) {
  std::string text = "";
  for (int i = 0; i < accent_phrases.size(); i++) {
    AccentPhraseModel phrase = accent_phrases[i];
    std::vector<MoraModel> moras = phrase.moras;
    for (int j = 0; j < moras.size(); j++) {
      MoraModel mora = moras[j];
      std::string vowel = mora.vowel;
      if (vowel == "A" || vowel == "I" || vowel == "U" || vowel == "E" || vowel == "O") {
        text += UNVOICE_SYMBOL;
      }
      text += mora.text;

      if (j + 1 == phrase.accent) {
        text += ACCENT_SYMBOL;
      }
    }

    if (phrase.is_interrogative) {
      text += WIDE_INTERROGATION_MARK;
    }

    if (i < accent_phrases.size()) {
      if (phrase.pause_mora != std::nullopt)
        text += PAUSE_DELIMITER;
      else
        text += NOPAUSE_DELIMITER;
    }
  }
  return text;
}
}  // namespace voicevox::core::engine

File: crates/voicevox_core/src/engine/kana_parser.rs
use crate::engine::model::{AccentPhraseModel, MoraModel};
use crate::engine::mora_list::MORA_LIST_MINIMUM;
use once_cell::sync::Lazy;
use std::collections::HashMap;

const UNVOICE_SYMBOL: char = '_';
const ACCENT_SYMBOL: char = '\'';
const NOPAUSE_DELIMITER: char = '/';
const PAUSE_DELIMITER: char = '、';
const WIDE_INTERROGATION_MARK: char = '？';
const LOOP_LIMIT: usize = 300;

#[derive(Clone, Debug, PartialEq, Eq, thiserror::Error)]
#[error("入力テキストをAquesTalk風記法としてパースすることに失敗しました: {_0}")]
pub(crate) struct KanaParseError(String);

type KanaParseResult<T> = std::result::Result<T, KanaParseError>;

static TEXT2MORA_WITH_UNVOICE: Lazy<HashMap<String, MoraModel>> = Lazy::new(|| {
    let mut text2mora_with_unvoice = HashMap::new();
    for [text, consonant, vowel] in MORA_LIST_MINIMUM {
        let consonant = if !consonant.is_empty() {
            Some(consonant.to_string())
        } else {
            None
        };
        let consonant_length = if consonant.is_some() { Some(0.0) } else { None };

        if ["a", "i", "u", "e", "o"].contains(vowel) {
            let upper_vowel = vowel.to_uppercase();
            let unvoice_mora = MoraModel::new(
                text.to_string(),
                consonant.clone(),
                consonant_length,
                upper_vowel,
                0.,
                0.,
            );
            text2mora_with_unvoice.insert(UNVOICE_SYMBOL.to_string() + text, unvoice_mora);
        }

        let mora = MoraModel::new(
            text.to_string(),
            consonant,
            consonant_length,
            vowel.to_string(),
            0.,
            0.,
        );
        text2mora_with_unvoice.insert(text.to_string(), mora);
    }
    text2mora_with_unvoice
});

fn text_to_accent_phrase(phrase: &str) -> KanaParseResult<AccentPhraseModel> {
    let phrase_vec: Vec<char> = phrase.chars().collect();
    let mut accent_index: Option<usize> = None;
    let mut moras: Vec<MoraModel> = Vec::new();
    let mut stack = String::new();
    let mut matched_text: Option<String> = None;
    let text2mora = &TEXT2MORA_WITH_UNVOICE;
    let mut index = 0;
    let mut loop_count = 0;
    while index < phrase_vec.len() {
        loop_count += 1;
        let letter = phrase_vec[index];
        if letter == ACCENT_SYMBOL {
            if index == 0 {
                return Err(KanaParseError(format!(
                    "accent cannot be set at beginning of accent phrase: {phrase}"
                )));
            }
            if accent_index.is_some() {
                return Err(KanaParseError(format!(
                    "second accent cannot be set at an accent phrase: {phrase}"
                )));
            }
            accent_index = Some(moras.len());
            index += 1;
            continue;
        }

        for &watch_letter in &phrase_vec[index..] {
            if watch_letter == ACCENT_SYMBOL {
                break;
            }
            stack.push(watch_letter);
            if text2mora.contains_key(&stack) {
                matched_text = Some(stack.clone());
            }
        }
        if let Some(matched_text) = matched_text.take() {
            index += matched_text.chars().count();
            moras.push(text2mora.get(&matched_text).unwrap().clone());
            stack.clear();
        } else {
            return Err(KanaParseError(format!(
                "unknown text in accent phrase: {phrase}"
            )));
        }
        if loop_count > LOOP_LIMIT {
            return Err(KanaParseError("detected infinity loop!".to_string()));
        }
    }
    if accent_index.is_none() {
        return Err(KanaParseError(format!(
            "accent not found in accent phrase: {phrase}"
        )));
    }
    Ok(AccentPhraseModel::new(
        moras,
        accent_index.unwrap(),
        None,
        false,
    ))
}

pub(crate) fn parse_kana(text: &str) -> KanaParseResult<Vec<AccentPhraseModel>> {
    const TERMINATOR: char = '\0';
    let mut parsed_result = Vec::new();
    let chars_of_text = text.chars().chain([TERMINATOR]);
    let mut phrase = String::new();
    for letter in chars_of_text {
        if letter == TERMINATOR || letter == PAUSE_DELIMITER || letter == NOPAUSE_DELIMITER {
            if phrase.is_empty() {
                return Err(KanaParseError(format!(
                    "accent phrase at position of {} is empty",
                    parsed_result.len()
                )));
            }
            let is_interrogative = phrase.contains(WIDE_INTERROGATION_MARK);
            if is_interrogative {
                if phrase.find(WIDE_INTERROGATION_MARK).unwrap()
                    != phrase.len() - WIDE_INTERROGATION_MARK.len_utf8()
                {
                    return Err(KanaParseError(format!(
                        "interrogative mark cannot be set at not end of accent phrase: {phrase}"
                    )));
                }
                phrase.pop(); // remove WIDE_INTERROGATION_MARK
            }
            let accent_phrase = {
                let mut accent_phrase = text_to_accent_phrase(&phrase)?;
                if letter == PAUSE_DELIMITER {
                    accent_phrase.set_pause_mora(Some(MoraModel::new(
                        PAUSE_DELIMITER.to_string(),
                        None,
                        None,
                        "pau".to_string(),
                        0.,
                        0.,
                    )));
                }
                accent_phrase.set_is_interrogative(is_interrogative);
                accent_phrase
            };
            parsed_result.push(accent_phrase);
            phrase.clear();
        } else {
            phrase.push(letter);
        }
    }
    Ok(parsed_result)
}

pub(crate) fn create_kana(accent_phrases: &[AccentPhraseModel]) -> String {
    let mut text = String::new();
    for phrase in accent_phrases {
        let moras = phrase.moras();
        for (index, mora) in moras.iter().enumerate() {
            if ["A", "E", "I", "O", "U"].contains(&(*mora.vowel()).as_ref()) {
                text.push(UNVOICE_SYMBOL);
            }
            text.push_str(mora.text());
            if index + 1 == *phrase.accent() {
                text.push(ACCENT_SYMBOL);
            }
        }
        if *phrase.is_interrogative() {
            text.push(WIDE_INTERROGATION_MARK);
        }
        text.push(if phrase.pause_mora().is_some() {
            PAUSE_DELIMITER
        } else {
            NOPAUSE_DELIMITER
        });
    }
    text.pop(); // remove last delimiter
    text
}

#[cfg(test)]
mod tests {
    use crate::engine::mora_list::MORA_LIST_MINIMUM;
    use pretty_assertions::assert_eq;
    use rstest::rstest;

    #[rstest]
    #[case(Some("da"), "ダ")]
    #[case(Some("N"), "ン")]
    #[case(Some("cl"), "ッ")]
    #[case(Some("sho"), "ショ")]
    #[case(Some("u"), "ウ")]
    #[case(Some("gA"), "_ガ")]
    #[case(Some("byO"), "_ビョ")]
    #[case(Some("O"), "_オ")]
    #[case(None, "fail")]
    fn test_text2mora_with_unvoice(#[case] mora: Option<&str>, #[case] text: &str) {
        let text2mora = &super::TEXT2MORA_WITH_UNVOICE;
        assert_eq!(text2mora.len(), MORA_LIST_MINIMUM.len() * 2 - 2); // added twice except ン and ッ
        let res = text2mora.get(text);
        assert_eq!(mora.is_some(), res.is_some());
        if let Some(res) = res {
            let mut m = String::new();
            if let Some(ref c) = *res.consonant() {
                m.push_str(c);
            }
            m.push_str(res.vowel());
            assert_eq!(m, mora.unwrap());
        }
    }

    #[rstest]
    #[case("ア_シタ'ワ", true)]
    #[case("ユウヒガ'", true)]
    #[case("_キ'レイ", true)]
    #[case("アクセントナシ", false)]
    #[case("アクセ'ント'タクサン'", false)]
    #[case("'アクセントハジマリ", false)]
    #[case("不明な'文字", false)]
    fn test_text_to_accent_phrase(#[case] text: &str, #[case] result_is_ok_expected: bool) {
        let result = super::text_to_accent_phrase(text);
        assert_eq!(result.is_ok(), result_is_ok_expected, "{:?}", result);
    }

    #[rstest]
    #[case("テ'ス_ト/テ_ス'ト、_テ'_スト？/テ'ス_ト？", true)]
    #[case("クウハクノ'//フレーズ'", false)]
    #[case("フレー？ズノ'/トチュウニ'、ギモ'ンフ", false)]
    fn test_parse_kana(#[case] text: &str, #[case] result_is_ok_expected: bool) {
        let result = super::parse_kana(text);
        assert_eq!(result.is_ok(), result_is_ok_expected, "{:?}", result);
    }
    #[rstest]
    fn test_create_kana() {
        let text = "アンドロ'イドワ、デンキ'/ヒ'_ツジノ/ユメ'オ/ミ'ルカ？";
        let phrases = super::parse_kana(text).unwrap();
        let text_created = super::create_kana(&phrases);
        assert_eq!(text, &text_created);
    }
}
